onchain/src/social_account.cairo

```diff
@@ -0,0 +1,88 @@
+use starknet::{ContractAddress, get_caller_address, get_contract_address, contract_address_const};
+
+
+#[starknet::interface]
+pub trait ISocialPayAccount<TContractState> {
+    fn get_public_key(self: @TContractState) -> u256;
+}
+
+
+#[starknet::contract]
+pub mod SocialPayAccount {
+    #[storage]
+    struct Storage {
+        #[key]
+        public_key: u256
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        AccountCreated: AccountCreated,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct AccountCreated {
+        #[key]
+        public_key: u256
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, public_key: u256) {
+        self.public_key.write(public_key);
+        self.emit(AccountCreated { public_key });
+    }
+
+    #[abi(embed_v0)]
+    impl SocialPayAccount of super::ISocialPayAccount<ContractState> {
+        fn get_public_key(self: @ContractState) -> u256 {
+            self.public_key.read()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use starknet::{
+        ContractAddress, get_caller_address, get_contract_address, contract_address_const
+    };
+    use snforge_std as snf;
+    use snforge_std::{
+        declare, ContractClassTrait, start_prank, stop_prank, CheatTarget, spy_events, SpyOn,
+        EventSpy, EventFetcher, Event
+    };
+    use super::{ISocialPayAccountDispatcher, ISocialPayAccountDispatcherTrait};
+
+
+    const public_key: u256 = 45;
+
+    fn deploy_social_account() -> ContractAddress {
+        let contract = declare("SocialPayAccount");
+
+        let mut social_account_calldata = array![];
+        public_key.serialize(ref social_account_calldata);
+
+        let address = contract.precalculate_address(@social_account_calldata);
+
+        let mut spy = spy_events(SpyOn::One(address));
+
+        let deployed_contract_address = contract.deploy(@social_account_calldata).unwrap();
+
+        spy.fetch_events();
+
+        assert(spy.events.len() == 1, 'there should be one event');
```

Comment:
Check that the emitted event public_key is correct.

Reply:
ok

Reply:
Am getting this crazy error

`Trait has no implementation in context: core::traits::Into::<@core::array::Array::<core::felt252>, core::integer::u256>`

This is some part of the code
```

        let (_, event) = spy.events.at(0);
        assert(event.keys.at(0) == @selector!("AccountCreated"), 'Wrong event name');

        assert(public_key == event.data.into(), 'Public key is not the same');
```

And am importing

use traits::Into;  // this also throw this error -> Identifier not found.


I want both public_key and  event.data.into() to in the same type u256

Reply:
from the documentation : https://foundry-rs.github.io/starknet-foundry/testing/testing-events.html

```
 let (_, event) = spy.events.at(0);
        assert(event.keys.at(0) == @selector!("AccountCreated"), 'Wrong event name');
```

we can only check the event name, we cannot check the event emitted values

Reply:
What about `assert_emitted`?

Reply:
i did not really understand what they did here 

```
 spy.assert_emitted(@array![
        (
            contract_address,
            SpyEventsChecker::Event::FirstEvent(
                SpyEventsChecker::FirstEvent { some_data: 123 }
            )
        )
    ]);
```

Reply:
does Into trait come with cairo core library by default

am getting the following error 

"Trait has no implementation in context: core::traits::Into::<@core::felt252, core::integer::u256>" 

```
for this code  let key_felt252 = event.keys.at(1);
        let key: u256 = key_felt252.into();
```

Reply:
deferencing the *event.keys.at(1);  fix the above error

```
   let key_felt252 = *event.keys.at(1);
        let key: u256 = key_felt252.into();
        assert(key == public_key, 'Wrong Public Key');
```